src/notification/types.js

```diff
@@ -0,0 +1,27 @@
+// @flow strict-local
+
+/**
+ * The data we need in JS/React code for acting on a notification.
+ *
+ * The actual objects we describe with these types may have a bunch more
+ * fields.  So, properly, these object types aren't really exact.  But
+ * pretending they are seems to be the least unpleasant way to get Flow to
+ * recognize the effect of refinements like `data.pm_users === undefined`.
+ *
+ * Currently:
+ *
+ *  * On iOS, these objects are translated directly, field by field, from
+ *    the blobs of key/value pairs sent by the Zulip server in the
+ *    "payload".  The set of fields therefore varies by server version.
```

Comment:
I think at the tip of the branch this jsdoc can be updated, to be made happier :grinning: 

(That's assuming that now all values we describe as `Notification`, in the iOS case, come from the new validating parser. Which I believe is a consequence of this branch, but haven't confirmed.)